### PR TITLE
Add: HiPACE

### PR DESCRIPTION
**Pull request by ax3l**
Add the new HiPACE++ code.

cc @MaxThevenet @SeverinDiederichs

## To Do

- [x] Linux testing @ax3l (OpenMP)
- [x] Linux testing (CUDA), macOS testing @SeverinDiederichs @MaxThevenet

### CODE REVIEW

File: var/spack/repos/builtin/packages/hipace/package.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    variant('precision',
+            default='double',
+            values=('single', 'double'),
+            multi=False,
+            description='Floating point precision (single/double)')
+
+    depends_on('cmake@3.15.0:', type='build')
+    depends_on('cuda@9.2.88:', when='compute=cuda')
+    depends_on('fftw@3:', when='+psatd compute=omp')
+    depends_on('fftw +mpi', when='+psatd +mpi compute=omp')
```
Comment (SeverinDiederichs) [inline var/spack/repos/builtin/packages/hipace/package.py]:
```suggestion
    depends_on('fftw@3:', when='compute=omp')
    depends_on('fftw +mpi', when='+mpi compute=omp')
```

File: var/spack/repos/builtin/packages/hipace/package.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    depends_on('mpi', when='+mpi')
+    depends_on('openpmd-api@hipace', when='+openpmd')
+    depends_on('openpmd-api +mpi', when='+openpmd +mpi')
+    depends_on('pkgconfig', type='build', when='+psatd compute=omp')
```
Comment (SeverinDiederichs) [inline var/spack/repos/builtin/packages/hipace/package.py]:
```suggestion
    depends_on('pkgconfig', type='build', when='compute=omp')
```

File: var/spack/repos/builtin/packages/hipace/package.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            multi=False,
+            description='Floating point precision (single/double)')
+
+    depends_on('cmake@3.15.0:', type='build')
+    depends_on('cuda@9.2.88:', when='compute=cuda')
+    depends_on('fftw@3:', when='compute=omp')
+    depends_on('fftw +mpi', when='+mpi compute=omp')
+    depends_on('mpi', when='+mpi')
+    depends_on('openpmd-api@hipace', when='+openpmd')
+    depends_on('openpmd-api +mpi', when='+openpmd +mpi')
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/hipace/package.py:44]:
```suggestion
    depends_on('openpmd-api ~mpi', when='+openpmd ~mpi')
    depends_on('openpmd-api +mpi', when='+openpmd +mpi')
```

**Review by SeverinDiederichs — APPROVED** [review]
I successfully tested both Linux + CUDA and macOS. Looks good!

**Review by haampie — APPROVED** [review]

**Comment by haampie** [comment]
Should it be a draft until it's tested?

**Comment by ax3l** [comment]
All tests successful, we are ready to :shipit: ship it :D

**Review by alalazo — APPROVED** [review]

**Pull request merged**